qiskit_experiments/framework/composite/composite_analysis.py

```diff
@@ -206,9 +207,24 @@ def _marginalized_component_data(self, composite_data: List[Dict]) -> List[List[
                         sub_data["counts"] = datum["counts"]
                 if "memory" in datum:
                     if composite_clbits is not None:
-                        sub_data["memory"] = (
-                            np.array(datum["memory"])[composite_clbits[i]]
-                        ).tolist()
+                        # level 2
+                        if isinstance(datum["memory"][0], str):
+                            num_cbits = 1 + max(
+                                cbit for cbit_list in composite_clbits for cbit in cbit_list
+                            )
+                            header = {"memory_slots": num_cbits}
+                            marginalized_memory = []
+                            for shot in datum["memory"]:
```

Comment:
Since ParallelExperiments always concatenate clbits for sub-experiments you are garuenteed that the composite_clbits are a continuous range. In this case you can simplify this processing like:
```python
clbits = composite_bits[i]
idx = slice(clbits[0], clbits[-1] + 1)
marginalized = [format_counts_memory(mem, header)[idx] for mem in datum["memory"]]
```

If performance becomes an issue for large numbers of shots (which im sure it will at some point) we should probably update or replace the `format_counts_memory` into a more efficient vectorized function that can go straight from list/array hex str -> list/array marginal counts

Reply:
Thanks. See here: [871400e](https://github.com/Qiskit/qiskit-experiments/pull/801/commits/871400e7da59d7a70322f0452932d93be2a3c26a) note that you still need the `[::-1]` to conform to Qiskit's bit ordering. Unless the `composite_bits` already account for this which I doubt? Note that before [871400e](https://github.com/Qiskit/qiskit-experiments/pull/801/commits/871400e7da59d7a70322f0452932d93be2a3c26a) I realized that we were calling `format_counts_memory` a total of `n_shots x n_subexperiments` times. This is inefficient and by pulling this out of the loop over `for i, index in enumerate(metadata["composite_index"])` we only need to call it `n_shots` times which is much better.